File: src/Java/1-100/96.UniqueBinarySearchTrees.java

import java.util.List;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashMap;
import java.util.Queue;
import java.util.HashSet;
import java.util.LinkedList;


public class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;
    TreeNode(int x) { val = x; }
}

class Solution {
    public int numTrees(int n) {
        int[] dp = new int[n + 1];
        dp[0] = 1;
        dp[1] = 1;
        // 卡特兰数
        for(int i = 2; i < n + 1; i++)
            for(int j = 1; j < i + 1; j++) 
                dp[i] += dp[j - 1] * dp[i - j];
        return dp[n];
    }
}

File: src/CSharp/96.UniqueBinarySearchTrees.cs

using System;
using System.Linq;
using System.Text;
using System.Collections.Generic;


namespace PeefyLeetCode.UniqueBinarySearchTrees
{

public class TreeNode {
    public int val;
    public TreeNode left;
    public TreeNode right;
    public TreeNode(int x) { val = x; }
}

public class Solution {
    public int NumTrees(int n) {
        if(n <= 1)
            return 1;
        long res = 1;
        for(int i = 1; i <= n; i++){
            res = res * (4 * i - 2) / (i + 1);
        }
        return (int) res;
    }
}
}